cpp/src/io/json/read_json.cu

```diff
@@ -278,25 +307,73 @@ table_with_metadata read_json(host_span<std::unique_ptr<datasource>> sources,
                  "Multiple inputs are supported only for JSON Lines format");
   }
 
-  datasource::owning_buffer<rmm::device_uvector<char>> bufview =
-    get_record_range_raw_input(sources, reader_opts, stream);
+  std::for_each(sources.begin(), sources.end(), [](const auto& source) {
+    CUDF_EXPECTS(source->size() < std::numeric_limits<int>::max(),
+                 "The size of each source file must be less than INT_MAX bytes");
+  });
 
-  // If input JSON buffer has single quotes and option to normalize single quotes is enabled,
-  // invoke pre-processing FST
-  if (reader_opts.is_enabled_normalize_single_quotes()) {
-    normalize_single_quotes(bufview, stream, rmm::mr::get_current_device_resource());
+  size_t const batch_size_ub = std::numeric_limits<int>::max();
+  size_t const chunk_offset  = reader_opts.get_byte_range_offset();
+  size_t chunk_size          = reader_opts.get_byte_range_size();
+  chunk_size                 = !chunk_size ? sources_size(sources, 0, 0) : chunk_size;
+
+  size_t start_source = 0, cur_size = 0;
+  for (size_t sum = 0; start_source < sources.size() && sum <= chunk_offset;
+       sum += sources[start_source]->size(), start_source++)
+    ;
+  start_source--;
+  std::vector<size_t> batch_positions, batch_sizes;
+  batch_positions.push_back(0);
+  for (size_t i = start_source; i < sources.size(); i++) {
+    cur_size += sources[i]->size();
+    if (cur_size >= batch_size_ub) {
+      batch_positions.push_back(i);
+      batch_sizes.push_back(cur_size - sources[i]->size());
+      cur_size = sources[i]->size();
+    }
   }
+  batch_positions.push_back(sources.size());
+  batch_sizes.push_back(cur_size);
 
-  // If input JSON buffer has unquoted spaces and tabs and option to normalize whitespaces is
-  // enabled, invoke pre-processing FST
-  if (reader_opts.is_enabled_normalize_whitespace()) {
-    normalize_whitespace(bufview, stream, rmm::mr::get_current_device_resource());
+  std::vector<cudf::io::table_with_metadata> partial_tables;
+  json_reader_options batched_reader_opts{reader_opts};
+
+  for (size_t i = 0; i < batch_sizes.size(); i++) {
+    batched_reader_opts.set_byte_range_size(std::min(batch_sizes[i], chunk_size));
+    partial_tables.emplace_back(read_batch(
+      host_span<std::unique_ptr<datasource>>(sources.begin() + batch_positions[i],
+                                             batch_positions[i + 1] - batch_positions[i]),
+      batched_reader_opts,
+      stream,
+      mr));
```

Comment:
we should not be using user provided mr here, since these tables won't be returned to the user. I think we might have to also handle the case where there's only one batch (concat not needed).

Reply:
Oh, which mr should be used instead?
Added an early return for the single batch case.

Reply:
`rmm::mr::get_current_device_resource()`